### DISCUSSION TITLE
Model Not Getting Any detections despite it training superbly.

### DISCUSSION THREAD

**Discussion by mwasela** [ultralytics/ultralytics#19646] [Q&A]
``I have trained a model based on YoloV8s for 250 Epochs using around 178 annotated images. I tried loading the models best.pt but getting no inferences at all, despite the training completing successfully. 

![WhatsApp Image 2025-03-11 at 19 45 38_a1a9ef36](https://github.com/user-attachments/assets/83ec7b2a-7436-4803-8206-9efa4d535cb4)
![WhatsApp Image 2025-03-11 at 19 52 57_d51822d1](https://github.com/user-attachments/assets/cafbd466-f29a-4889-89e6-19eca90e3919)
![WhatsApp Image 2025-03-11 at 19 53 38_24c2f397](https://github.com/user-attachments/assets/0119ce1c-d0a2-409a-99d2-19bf9001622f)
![WhatsApp Image 2025-03-11 at 19 53 39_4b0d1267](https://github.com/user-attachments/assets/6bf63143-2400-4526-8a4a-2f991e05bad6)
![WhatsApp Image 2025-03-11 at 19 53 39_356a20d4](https://github.com/user-attachments/assets/435f2e4c-e7e0-4bb9-b5a9-4141fddcae83)
![WhatsApp Image 2025-03-11 at 19 53 39_a3d48310](https://github.com/user-attachments/assets/bc7a4c6e-944f-4115-9aa7-57d75ddde830)

`
import cv2
from ultralytics import YOLO

# Load the trained YOLOv8 model
model = YOLO('best.pt')  # Path to the trained YOLOv8 model 

# RTSP feed or video file URL
video_source = "output.mp4"  # Ensure this is a valid path

# Create a VideoCapture object
cap = cv2.VideoCapture(video_source)

if not cap.isOpened():
    print("Error: Unable to open the video stream")
    exit()

# Read frames from the RTSP feed
while True:
    ret, frame = cap.read()
    if not ret:
        print("Failed to grab frame")
        break

    # Resize frame for YOLO model input (640x640 recommended)
    frame_resized = cv2.resize(frame, (640, 640))  # Resize to 640x640

    # Inference on the frame with lower confidence threshold
    results = model(frame_resized, conf=0.25)

    # Check the results
    print(results)  # Debugging: Print the results to check predictions

    # Get annotated frame (visualize the predictions)
    annotated_frame = results[0].plot()

    # Display the frame with detections
    cv2.imshow("YOLOv8 Inference", annotated_frame)

    # Press 'q' to quit
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release resources
cap.release()
cv2.destroyAllWindows()

`

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mwasela, thank you for sharing your concern and for your interest in Ultralytics 🚀! 

To assist you better, could you please provide a [minimum reproducible example (MRE)] to help us debug the issue? An MRE ensures that we can isolate and identify the problem effectively. Additionally, it would be helpful to know if you’ve verified your dataset annotations, confirmed your model's training parameters, and checked if the `best.pt` file is being loaded correctly.

If this is related to a deployment or inference issue, you might want to double-check your input preprocessing (e.g., resizing images to the correct dimensions) and ensure your confidence threshold during inference is appropriate. 

An Ultralytics engineer will review your query and provide further assistance soon. In the meantime, feel free to share more details about your setup, such as your training logs, dataset examples, and any additional error messages you might have encountered. Stay tuned for further support! 😊

**Comment by mwasela** [comment]
Ive tried putting confidence thresholds but no results. I can share the model best.pt file link here if someone can try testing. Any video with bags in it will do. Find the last.pt, best.pt and a sample video for anyone who wants to test. I have a similar model I trained before, when I plug it in it works fine. I don't get why this new one does not work. [Link to files](https://drive.google.com/drive/folders/1AsPQ-flAxsMYHo7l6XlkT9gbjS5D1wgr?usp=sharing)

**Reply by glenn-jocher** [reply]
@mwasela thanks for sharing the details. Let's help troubleshoot this step by step. First, could you verify two key points from your training setup: 

1. **Dataset Configuration**: Ensure your YAML file correctly maps class names to IDs (0 for 'bag' in your case). The training images you shared show annotations, but we need to confirm the labels match the data format expected by YOLOv8.

2. **Inference Pipeline**: Instead of manual resizing, let the model handle preprocessing by using `model.predict(source=frame)` directly on the original frame. Your current resizing to 640x640 might be introducing aspect ratio distortion that's affecting detections.

For immediate debugging, try running inference with: 
```python
results = model.predict(frame, imgsz=640, conf=0.25)
```
and skip the manual resizing step. If issues persist, you might want to validate your model using the Val mode as shown in our Object Detection with YOLOv8 guide. Let us know how these steps work!

**Comment by mwasela** [comment]
So this was my config.yaml
```
path: C:/Users/lab/Documents/model_two/data
train: images/
val: images/

# Classes
nc: 1
names:
  0: 'Bag'
```

The training was  also able to generate some labelled detections,  let me
share.

[image: val_batch0_labels.jpg][image: val_batch0_pred.jpg]

See, and when I load the model and console log the prediction, it does say
model class 0:bag.

But let me try the above two suggestions as you go through this.

On Tue, Mar 11, 2025 at 4:13 PM Glenn Jocher ***@***.***>
wrote:

> @mwasela <https://github.com/mwasela> thanks for sharing the details.
> Let's help troubleshoot this step by step. First, could you verify two key
> points from your training setup:
>
>    1.
>
>    *Dataset Configuration*: Ensure your YAML file correctly maps class
>    names to IDs (0 for 'bag' in your case). The training images you shared
>    show annotations, but we need to confirm the labels match the data format
>    expected by YOLOv8.
>    2.
>
>    *Inference Pipeline*: Instead of manual resizing, let the model handle
>    preprocessing by using model.predict(source=frame) directly on the
>    original frame. Your current resizing to 640x640 might be introducing
>    aspect ratio distortion that's affecting detections.
>
> For immediate debugging, try running inference with:
>
> results = model.predict(frame, imgsz=640, conf=0.25)
>
> and skip the manual resizing step. If issues persist, you might want to
> validate your model using the Val mode as shown in our Object Detection
> with YOLOv8 guide. Let us know how these steps work!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/ultralytics/discussions/19646#discussioncomment-12467815>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACXRGWAP6DKCHAG4W7TRBGT2T5U3FAVCNFSM6AAAAABYZSGUNOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENBWG44DCNI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 

Regards

*Dennis Michael Mwasela*
*Research & Development **Lead *
*Microdam Telecommunications LLC*
*Automation Design Engineers & Consultants*

Industrial IoT | Robotics | Solar Power | Engineering Consultants | Plant
Mechanics | R & D

Alt Email: ***@***.***
Work Phone: +254 717217016
https://microdam.co.ke

**Reply by glenn-jocher** [reply]
@mwasela thanks for sharing the additional details. A few quick suggestions to resolve the detection issues:

1. **Dataset Split**: Having the same `train:` and `val:` paths in your YAML (both pointing to `images/`) likely caused overfitting. Create separate train/val splits (80/20 recommended) as shown in the Object Detection Datasets guide.

2. **BGR/RGB Conversion**: Add `frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)` before inference since OpenCV reads frames in BGR format while YOLO expects RGB.

3. **Direct Validation**: Test your model on the same validation images used during training using:  
```python
model.val(data='your_data.yaml')
```
This will confirm if the issue is inference-related or a broader model problem.

Let us know if these adjustments help surface detections!

**Comment by Y-T-G** [comment]
Follow the Ultralytics docs to make prediction instead of AI-generated code that's incorrect.

https://docs.ultralytics.com/modes/predict/

Also that training graph looks like overfitting, which is typical with so few images. You can't rely on the metrics if you have so few images. They may as well just be fluke.

You need a lot more images for a model that works well.

https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/
